yt/yt/server/lib/chunk_pools/ordered_chunk_pool.cpp

```diff
@@ -404,26 +408,155 @@ class TOrderedChunkPool
             : JobSizeConstraints_->GetDataWeightPerJob();
     }
 
-    IChunkPoolOutput::TCookie AddPrimaryDataSlice(
+    i64 GetCurrentJobDataWeight()
+    {
+        return JobCarryOverDataWeight_ + CurrentJob()->GetDataWeight();
+    }
+
+    IChunkPoolOutput::TCookie AddUnsplittablePrimaryDataSlice(
         const TLegacyDataSlicePtr& dataSlice,
         IChunkPoolInput::TCookie cookie,
         i64 dataSizePerJob)
     {
         YT_VERIFY(!dataSlice->IsLegacy);
-        bool jobIsLargeEnough =
-            CurrentJob()->GetPreliminarySliceCount() + 1 > JobSizeConstraints_->GetMaxDataSlicesPerJob() ||
-            CurrentJob()->GetDataWeight() >= dataSizePerJob;
+
+        RowCountUntilJobSplit_ = 0;
+        JobCarryOverDataWeight_ = 0;
+
         IChunkPoolOutput::TCookie result = IChunkPoolOutput::NullCookie;
+
+        bool jobIsLargeEnough =
+            CurrentJob()->GetPreliminarySliceCount() >= JobSizeConstraints_->GetMaxDataSlicesPerJob() ||
+            GetCurrentJobDataWeight() >= dataSizePerJob;
         if (jobIsLargeEnough && !SingleJob_) {
             result = EndJob();
         }
+
         auto dataSliceCopy = CreateInputDataSlice(dataSlice);
         dataSliceCopy->Tag = cookie;
-        dataSlice->CopyPayloadFrom(*dataSlice);
-        CurrentJob()->AddDataSlice(dataSliceCopy, cookie, true /*isPrimary*/);
+        CurrentJob()->AddDataSlice(dataSliceCopy, cookie, /*isPrimary*/ true);
         return result;
     }
 
+    bool IsDataSliceSplittable(const TLegacyDataSlicePtr& dataSlice) const
```

Comment:
nip: Also maybe IsSplittable?

Reply:
Such a method already exists (and it's virtual) :(